// @pm2/io/build/main/src/actions/coverage.js
"use strict";

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var debug_1 = require("debug");
var debug = debug_1.default('axm:coverageaction');
var coverage_1 = require("../features/coverage");
var miscellaneous_1 = require("../utils/miscellaneous");
var serviceManager_1 = require("../serviceManager");

var CoverageAction = function () {
    function CoverageAction(actionFeature) {
        _classCallCheck(this, CoverageAction);

        this.actionFeature = actionFeature;
    }

    _createClass(CoverageAction, [{
        key: "init",
        value: function init() {
            return tslib_1.__awaiter(this, void 0, void 0, /*#__PURE__*/regeneratorRuntime.mark(function _callee() {
                return regeneratorRuntime.wrap(function _callee$(_context) {
                    while (1) {
                        switch (_context.prev = _context.next) {
                            case 0:
                                this.coverageFeature = new coverage_1.default();
                                _context.prev = 1;
                                _context.next = 4;
                                return this.coverageFeature.init();

                            case 4:
                                // expose actions only if the feature is available
                                this.exposeActions();
                                _context.next = 10;
                                break;

                            case 7:
                                _context.prev = 7;
                                _context.t0 = _context["catch"](1);

                                debug("Failed to load coverage profiler: " + _context.t0.message);

                            case 10:
                            case "end":
                                return _context.stop();
                        }
                    }
                }, _callee, this, [[1, 7]]);
            }));
        }
    }, {
        key: "destroy",
        value: function destroy() {
            if (this.coverageFeature) this.coverageFeature.destroy();
        }
    }, {
        key: "stopCoverage",
        value: function stopCoverage(reply) {
            return tslib_1.__awaiter(this, void 0, void 0, /*#__PURE__*/regeneratorRuntime.mark(function _callee2() {
                var dumpFile;
                return regeneratorRuntime.wrap(function _callee2$(_context2) {
                    while (1) {
                        switch (_context2.prev = _context2.next) {
                            case 0:
                                _context2.prev = 0;
                                _context2.next = 3;
                                return this.coverageFeature.stop();

                            case 3:
                                dumpFile = _context2.sent;
                                return _context2.abrupt("return", reply({
                                    success: true,
                                    coverage: true,
                                    dump_file: dumpFile,
                                    dump_file_size: dumpFile.length,
                                    uuid: this.uuid
                                }));

                            case 7:
                                _context2.prev = 7;
                                _context2.t0 = _context2["catch"](0);
                                return _context2.abrupt("return", reply({
                                    success: false,
                                    err: _context2.t0,
                                    uuid: this.uuid
                                }));

                            case 10:
                            case "end":
                                return _context2.stop();
                        }
                    }
                }, _callee2, this, [[0, 7]]);
            }));
        }
    }, {
        key: "exposeActions",
        value: function exposeActions() {
            var _this = this;

            this.actionFeature.action('km:coverage:start', function (opts, reply) {
                return tslib_1.__awaiter(_this, void 0, void 0, /*#__PURE__*/regeneratorRuntime.mark(function _callee4() {
                    var _this2 = this;

                    return regeneratorRuntime.wrap(function _callee4$(_context4) {
                        while (1) {
                            switch (_context4.prev = _context4.next) {
                                case 0:
                                    if (!reply) {
                                        reply = opts;
                                        opts = {};
                                    }
                                    _context4.prev = 1;

                                    this.uuid = miscellaneous_1.default.generateUUID();
                                    _context4.next = 5;
                                    return this.coverageFeature.start(opts);

                                case 5:
                                    reply({ success: true, uuid: this.uuid });
                                    if (opts.timeout && typeof opts.timeout === 'number') {
                                        setTimeout(function (_) {
                                            return tslib_1.__awaiter(_this2, void 0, void 0, /*#__PURE__*/regeneratorRuntime.mark(function _callee3() {
                                                var reply;
                                                return regeneratorRuntime.wrap(function _callee3$(_context3) {
                                                    while (1) {
                                                        switch (_context3.prev = _context3.next) {
                                                            case 0:
                                                                reply = function reply(data) {
                                                                    return serviceManager_1.ServiceManager.get('transport').send('axm:reply', {
                                                                        return: data,
                                                                        action_name: 'km:coverage:stop'
                                                                    });
                                                                };

                                                                _context3.next = 3;
                                                                return this.stopCoverage(reply);

                                                            case 3:
                                                            case "end":
                                                                return _context3.stop();
                                                        }
                                                    }
                                                }, _callee3, this);
                                            }));
                                        }, opts.timeout);
                                    }
                                    _context4.next = 12;
                                    break;

                                case 9:
                                    _context4.prev = 9;
                                    _context4.t0 = _context4["catch"](1);
                                    return _context4.abrupt("return", reply({
                                        success: false,
                                        err: _context4.t0,
                                        uuid: this.uuid
                                    }));

                                case 12:
                                case "end":
                                    return _context4.stop();
                            }
                        }
                    }, _callee4, this, [[1, 9]]);
                }));
            });
            this.actionFeature.action('km:coverage:stop', this.stopCoverage.bind(this));
        }
    }]);

    return CoverageAction;
}();

exports.default = CoverageAction;
